{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Software info**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "|Software     |Version|\n",
    "|-------------|-------|\n",
    "|python    |3.11.9|\n",
    "|ipykernel    |[6.28.0](https://anaconda.org/anaconda/ipykernel)|\n",
    "|Biopython    |[1.70](https://anaconda.org/bioconda/biopython)|\n",
    "|Entrez-direct|[21.6](https://anaconda.org/bioconda/entrez-direct)|\n",
    "|mafft        |[7.525](https://anaconda.org/bioconda/mafft)|\n",
    "|iq-tree2     |[2.3.0](https://anaconda.org/bioconda/iqtree)|\n",
    "|DSTU         |[0.5.0 pre-release](https://github.com/iliapopov17/Detailed-Sequences-for-Trees-Unblemished)|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Conda envinronment: `dstu_hantavirus_phylo.yaml`<br>\n",
    "Install the envinronment with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "! conda env create -f dstu_hantavirus_phylo.yaml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Reload VS Code (close & open), then activate this envinronment as kernel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Hardware info**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- OS: Ubuntu 22.04 (Windows Subsystem for Linux)\n",
    "- CPU: Intel Xeon E5-2670v3\n",
    "- RAM: 32GB (16GB for WSL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Architecture:            x86_64\n",
      "  CPU op-mode(s):        32-bit, 64-bit\n",
      "  Address sizes:         46 bits physical, 48 bits virtual\n",
      "  Byte Order:            Little Endian\n",
      "CPU(s):                  24\n",
      "  On-line CPU(s) list:   0-23\n",
      "Vendor ID:               GenuineIntel\n",
      "  Model name:            Intel(R) Xeon(R) CPU E5-2670 v3 @ 2.30GHz\n",
      "    CPU family:          6\n",
      "    Model:               63\n",
      "    Thread(s) per core:  2\n",
      "    Core(s) per socket:  12\n",
      "    Socket(s):           1\n",
      "    Stepping:            2\n",
      "    BogoMIPS:            4589.36\n",
      "    Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mc\n",
      "                         a cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscal\n",
      "                         l nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopo\n",
      "                         logy cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1\n",
      "                          sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervis\n",
      "                         or lahf_lm abm invpcid_single pti ssbd ibrs ibpb stibp \n",
      "                         tpr_shadow vnmi ept vpid ept_ad fsgsbase bmi1 avx2 smep\n",
      "                          bmi2 erms invpcid xsaveopt md_clear flush_l1d arch_cap\n",
      "                         abilities\n",
      "Virtualization features: \n",
      "  Virtualization:        VT-x\n",
      "  Hypervisor vendor:     Microsoft\n",
      "  Virtualization type:   full\n",
      "Caches (sum of all):     \n",
      "  L1d:                   384 KiB (12 instances)\n",
      "  L1i:                   384 KiB (12 instances)\n",
      "  L2:                    3 MiB (12 instances)\n",
      "  L3:                    30 MiB (1 instance)\n",
      "Vulnerabilities:         \n",
      "  Itlb multihit:         KVM: Mitigation: VMX disabled\n",
      "  L1tf:                  Mitigation; PTE Inversion; VMX conditional cache flushe\n",
      "                         s, SMT vulnerable\n",
      "  Mds:                   Mitigation; Clear CPU buffers; SMT Host state unknown\n",
      "  Meltdown:              Mitigation; PTI\n",
      "  Spec store bypass:     Mitigation; Speculative Store Bypass disabled via prctl\n",
      "                          and seccomp\n",
      "  Spectre v1:            Mitigation; usercopy/swapgs barriers and __user pointer\n",
      "                          sanitization\n",
      "  Spectre v2:            Mitigation; Full generic retpoline, IBPB conditional, I\n",
      "                         BRS_FW, STIBP conditional, RSB filling\n",
      "  Srbds:                 Not affected\n",
      "  Tsx async abort:       Not affected\n"
     ]
    }
   ],
   "source": [
    "! lscpu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 0. Install `DSTU`**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2024-04-26 13:56:52--  https://github.com/iliapopov17/Detailed-Sequences-for-Trees-Unblemished/releases/download/v0.5.0-alpha/DSTU.py\n",
      "Resolving github.com (github.com)... 140.82.121.4\n",
      "Connecting to github.com (github.com)|140.82.121.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/791951883/7f3cedb0-f7f1-4f98-84a1-1c30f0cd382f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240426%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240426T105631Z&X-Amz-Expires=300&X-Amz-Signature=18504b055951147c15327de034f8197f475cbf5ebe89f161e5a9a93fdae9e3df&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=791951883&response-content-disposition=attachment%3B%20filename%3DDSTU.py&response-content-type=application%2Foctet-stream [following]\n",
      "--2024-04-26 13:56:52--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/791951883/7f3cedb0-f7f1-4f98-84a1-1c30f0cd382f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240426%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240426T105631Z&X-Amz-Expires=300&X-Amz-Signature=18504b055951147c15327de034f8197f475cbf5ebe89f161e5a9a93fdae9e3df&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=791951883&response-content-disposition=attachment%3B%20filename%3DDSTU.py&response-content-type=application%2Foctet-stream\n",
      "Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...\n",
      "Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.109.133|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 14405 (14K) [application/octet-stream]\n",
      "Saving to: ‘DSTU.py’\n",
      "\n",
      "DSTU.py             100%[===================>]  14.07K  --.-KB/s    in 0.004s  \n",
      "\n",
      "2024-04-26 13:56:53 (3.12 MB/s) - ‘DSTU.py’ saved [14405/14405]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget https://github.com/iliapopov17/Detailed-Sequences-for-Trees-Unblemished/releases/download/v0.5.0-alpha/DSTU.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from DSTU import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 1. Download sequences**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`accession_numbers.txt` file was created manually based on previously published papers:\n",
    "1. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10025241/<br>\n",
    "2. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7106157/<br>\n",
    "3. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10943075/<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded: NC_038515\n",
      "Downloaded: KT316176\n",
      "Downloaded: MN337866\n",
      "Downloaded: MG663536\n",
      "Downloaded: JX193700\n",
      "Downloaded: KX779125\n",
      "Downloaded: KU950715\n",
      "Downloaded: KR920360\n",
      "Downloaded: OM912841\n",
      "Downloaded: KX845680\n",
      "Downloaded: OM912842\n",
      "Downloaded: OM912840\n",
      "Downloaded: OM963009\n",
      "Downloaded: OM912844\n",
      "Downloaded: OM912843\n",
      "Downloaded: MK165653\n",
      "Downloaded: KJ000540\n",
      "Downloaded: KY040508\n",
      "Downloaded: KT899703\n",
      "Downloaded: MN850095\n",
      "Downloaded: AF005729\n",
      "Downloaded: NC_043407\n",
      "Downloaded: GQ200821\n",
      "Downloaded: EU788002\n",
      "Downloaded: GU997097\n",
      "Downloaded: FJ858378\n",
      "Downloaded: NC_038299\n",
      "Downloaded: AB620102\n",
      "Downloaded: AB620105\n",
      "Downloaded: HM015222\n",
      "Downloaded: MN639746\n",
      "Downloaded: AB677488\n",
      "Downloaded: JX028271\n",
      "Downloaded: KJ857315\n",
      "Downloaded: KJ857316\n",
      "Downloaded: MK883761\n",
      "Downloaded: KY751731\n",
      "Downloaded: MK542664\n",
      "Downloaded: MT441741\n",
      "Downloaded: HQ728461\n",
      "Downloaded: KC880348\n",
      "Downloaded: GU566021\n",
      "Downloaded: KJ857320\n",
      "Downloaded: KF974361\n",
      "Downloaded: JX990941\n",
      "Downloaded: MN006903\n",
      "Downloaded: JX990965\n",
      "Downloaded: KF705679\n",
      "Downloaded: EU929078\n",
      "Downloaded: NC_043175\n",
      "Downloaded: EF050454\n",
      "Downloaded: KC631784\n",
      "Downloaded: HQ834697\n",
      "Downloaded: MF683846\n",
      "Downloaded: KC490923\n",
      "Downloaded: AM998806\n",
      "Downloaded: LC553715\n",
      "Downloaded: GU904045\n",
      "Downloaded: AB620030\n",
      "Downloaded: AY675354\n",
      "Downloaded: MN608087\n",
      "Downloaded: DQ268652\n",
      "Downloaded: KF704716\n",
      "Downloaded: KF704717\n",
      "Downloaded: KF704715\n",
      "Downloaded: JX912953\n",
      "Downloaded: KF704713\n",
      "Downloaded: JN037851\n",
      "Downloaded: KT316176\n",
      "Downloaded: KM102249\n",
      "Downloaded: JX990946\n",
      "Downloaded: KR072623\n",
      "Downloaded: KF031355\n",
      "Downloaded: FJ593498\n",
      "Downloaded: KF010517\n",
      "Downloaded: KU950715\n",
      "Downloaded: JQ287716\n",
      "Downloaded: JX465369\n",
      "Downloaded: EU424341\n",
      "Downloaded: EF641806\n",
      "Downloaded: EF641807\n",
      "Downloaded: JX193700\n",
      "Downloaded: EU001330\n",
      "Downloaded: JX193697\n",
      "Downloaded: JX465379\n",
      "Downloaded: JX465383\n",
      "Downloaded: JX465381\n",
      "Downloaded: JX465382\n",
      "Downloaded: OR684449\n",
      "Downloaded: MG663536\n",
      "Downloaded: NC_038515\n",
      "Downloaded: NC_078262\n",
      "Downloaded: NC_055632\n",
      "Downloaded: NC_034401\n",
      "Downloaded: FJ593498\n",
      "Downloaded: KX512432\n",
      "Downloaded: NC_078485\n",
      "Downloaded: OR934973\n",
      "Downloaded: OR934974\n",
      "All downloads completed.\n"
     ]
    }
   ],
   "source": [
    "get_sequences('iljapopov17@gmail.com', 'data/accession_numbers.txt', 'genbank_sequences')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 1.1. Check downloaded sequences**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "93\n"
     ]
    }
   ],
   "source": [
    "! ls genbank_sequences/| wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The number of accession numbers is 99, but there are 93 downloaded sequences. There are non unique accession numbers extracted from one of these papers:\n",
    "1. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10025241/<br>\n",
    "2. https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7106157/<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_non_unique_strings(file_path):\n",
    "    counts = {}\n",
    "    non_unique_count = 0\n",
    "    with open(file_path, 'r') as file:\n",
    "        for line in file:\n",
    "            line = line.strip()\n",
    "            if line in counts:\n",
    "                if counts[line] == 1:\n",
    "                    non_unique_count += 1\n",
    "                counts[line] += 1\n",
    "            else:\n",
    "                counts[line] = 1\n",
    "\n",
    "    return non_unique_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of non-unique strings: 6\n"
     ]
    }
   ],
   "source": [
    "file = 'data/accession_numbers.txt'\n",
    "non_unique = count_non_unique_strings(file)\n",
    "print(f\"Number of non-unique strings: {non_unique}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Above mentioned papers utilize 6 identical sequences for building the phylogenetic tree. 93 sequences downloaded is explained in that case. No data has been lost."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 1.2. 10 additional sequences**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "10 additional hantaviruses associated with human infections will be included in our analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Orthohantavirus seoulense**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/MG386252.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/29774860/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"MG386252\" | efetch -format fasta > genbank_sequences/MG386252.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"MG386252\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/OR047284.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/38147030/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"OR047284\" | efetch -format fasta > genbank_sequences/OR047284.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"OR047284\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Orthohantavirus tulaense**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/KU297981.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/26691901/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"KU297981\" | efetch -format fasta > genbank_sequences/KU297981.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"KU297981\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/MT993951.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/33754997/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"MT993951\" | efetch -format fasta > genbank_sequences/MT993951.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"MT993951\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Orthohantavirus dobravaense**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/MK605664.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/31625853/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"MK605664\" | efetch -format fasta > genbank_sequences/MK605664.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"MK605664\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/MK605665.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/31625853/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"MK605665\" | efetch -format fasta > genbank_sequences/MK605665.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"MK605665\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Hantaan orthohantavirus**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/MW349026.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/34370707/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"MW349026\" | efetch -format fasta > genbank_sequences/MW349026.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"MW349026\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/MZ191082.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/34370707/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"MZ191082\" | efetch -format fasta > genbank_sequences/MZ191082.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"MZ191082\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Orthohantavirus sinnombreense**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/ON571586.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/37486231/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"ON571586\" | efetch -format fasta > genbank_sequences/ON571586.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"ON571586\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/nuccore/ON571589.1<br>\n",
    "https://pubmed.ncbi.nlm.nih.gov/37486231/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "! esearch -db nucleotide -query \"ON571589\" | efetch -format fasta > genbank_sequences/ON571589.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "! echo \"ON571589\" >> data/accession_numbers.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 2. Combine all sequences to one file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "! cat genbank_sequences/*.fasta > all_seqs.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The number of sequences in combined file: 103\n"
     ]
    }
   ],
   "source": [
    "with open(\"all_seqs.fa\", \"r\") as fasta_file:\n",
    "    content = fasta_file.read()\n",
    "    num_sequences = content.count(\">\")\n",
    "print(f\"The number of sequences in combined file: {num_sequences}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 3. Multiple sequences alignment**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      "\n",
      "There are 20 ambiguous characters.\n",
      "  101 / 103\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      "  100 / 103\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP    38 / 102 \n",
      "Reallocating..done. *alloclen = 14242\n",
      "STEP   102 / 102 \n",
      "done.\n",
      "\n",
      "Making a distance matrix from msa.. \n",
      "  100 / 103\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=1) ... \n",
      "  100 / 103\n",
      "done.\n",
      "\n",
      "Progressive alignment 2/2... \n",
      "STEP    91 / 102 \n",
      "Reallocating..done. *alloclen = 14215\n",
      "STEP   102 / 102 \n",
      "done.\n",
      "\n",
      "disttbfast (nuc) Version 7.505\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "dndpre (nuc) Version 7.505\n",
      "alg=X, model=DNA200 (2), 1.53 (4.59), 0.37 (1.11), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "minimumweight = 0.000010\n",
      "autosubalignment = 0.000000\n",
      "nthread = 0\n",
      "randomseed = 0\n",
      "blosum 62 / kimura 200\n",
      "poffset = 0\n",
      "niter = 2\n",
      "sueff_global = 0.100000\n",
      "nadd = 2\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "\n",
      "  100 / 103\n",
      "Segment   1/ 12    1-3078\n",
      "Segment   2/ 12 3030-3033.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    accepted. identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    accepted. identical.    identical.    rejected. identical.    accepted. rejected. rejected. identical.    identical.    accepted. identical.    identical.    accepted. identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. rejected. accepted. accepted. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    accepted. identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    accepted. identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    rejected. rejected. identical.    identical.    identical.    accepted. accepted. identical.    identical.    accepted. identical.    identical.    accepted. rejected. rejected. rejected. accepted. identical.    rejected. rejected. identical.    identical.    identical.    identical.    identical.    identical.    rejected. accepted. identical.    accepted. identical.    accepted. accepted. accepted. identical.    identical.    identical.    accepted. identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    accepted. identical.    accepted. identical.    identical.    accepted. identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    rejected. identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    rejected. accepted. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    accepted. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. accepted. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    rejected. identical.    identical.    rejected. identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.   \n",
      "STEP 002-101-1  identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   3/ 12 3033-3041\n",
      "STEP 002-101-1  identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   4/ 12 3041-3114\n",
      "STEP 002-101-1  identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   5/ 12 3114-3123\n",
      "STEP 002-101-1  identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   6/ 12 3123-3128\n",
      "STEP 002-101-1  identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   7/ 12 3128-3146\n",
      "STEP 002-101-1  identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   8/ 12 3146-3285\n",
      "STEP 002-002-0  identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment   9/ 12 3285-3365\n",
      "STEP 002-101-1  identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  10/ 12 3365-3374\n",
      "STEP 002-039-0  identical.    identical.    identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  11/ 12 3374-3378\n",
      "STEP 002-101-1  identical.    identical.    identical.    identical.    identical.   \n",
      "Converged.\n",
      "\n",
      "Segment  12/ 12 3378-6732\n",
      "done 002-001-1  identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    accepted. accepted. identical.    identical.    identical.    identical.    identical.    rejected. rejected. identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    rejected. identical.    rejected. accepted. accepted. identical.    rejected. identical.    accepted. identical.    identical.    rejected. rejected. identical.    rejected. rejected. rejected. rejected. rejected. identical.    identical.    rejected. identical.    rejected. identical.    rejected. rejected. rejected. identical.    rejected. identical.    rejected. identical.    rejected. identical.    identical.    identical.    rejected. identical.    identical.    accepted. rejected. identical.    rejected. accepted. identical.    rejected. accepted. identical.    identical.    identical.    identical.    rejected. rejected. identical.    identical.    identical.    identical.    accepted. accepted. rejected. accepted. identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    accepted. identical.    identical.    rejected. identical.    rejected. rejected. accepted. identical.    accepted. rejected. accepted. accepted. accepted. identical.    identical.    rejected. rejected. identical.    identical.    accepted. identical.    identical.    accepted. rejected. identical.    accepted. accepted. rejected. accepted. accepted. accepted. identical.    rejected. identical.    rejected. rejected. identical.    identical.    rejected. accepted. rejected. accepted. rejected. rejected. rejected. identical.    rejected. rejected. rejected. rejected. rejected. rejected. rejected. identical.    rejected. rejected. rejected. identical.    rejected. rejected. identical.    identical.    accepted. rejected. rejected. identical.    rejected. identical.    accepted. rejected. accepted. identical.    rejected. rejected. identical.    rejected. identical.    identical.    identical.    accepted. accepted. accepted. identical.    identical.    identical.    identical.    identical.    rejected. accepted. accepted. accepted. accepted. accepted. identical.    identical.    rejected. identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    rejected. accepted. rejected. accepted. identical.    identical.    rejected. identical.    rejected. rejected. rejected. accepted. accepted. identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    rejected. rejected. identical.    identical.    identical.    accepted. rejected. identical.    rejected. rejected. accepted. rejected. rejected. identical.    rejected. rejected. rejected. identical.    identical.    identical.    identical.    identical.    rejected. identical.    identical.    identical.    rejected. identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.    identical.   \n",
      "dvtditr (nuc) Version 7.505\n",
      "alg=A, model=DNA200 (2), 1.53 (4.59), -0.00 (-0.00), noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "Strategy:\n",
      " FFT-NS-i (Standard)\n",
      " Iterative refinement method (max. 2 iterations)\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! mafft --auto data/all_seqs.fa > data/all_seqs_mafft.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 4. Launching `ModelFinder` to get the best substitution model**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    IliaPC (AVX2, FMA3, 15 GB RAM)\n",
      "Command: iqtree2 -m MFP -s all_seqs_mafft.fa --prefix tree_MF2 -T AUTO\n",
      "Seed:    51880 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Sat Apr 20 14:36:56 2024\n",
      "Kernel:  AVX+FMA - auto-detect threads (24 CPU cores detected)\n",
      "\n",
      "Reading alignment file all_seqs_mafft.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 103 sequences with 6727 columns, 5018 distinct patterns\n",
      "4543 parsimony-informative, 346 singleton sites, 1838 constant sites\n",
      "             Gap/Ambiguity  Composition  p-value\n",
      "   1  AB620030.1     3.46%    passed     70.90%\n",
      "   2  AB620102.1     2.94%    passed      9.29%\n",
      "   3  AB620105.1     3.12%    passed     42.74%\n",
      "   4  AB677488.1     5.11%    passed     11.73%\n",
      "   5  AF005729.1    92.26%    passed     59.80%\n",
      "   6  AM998806.1    93.88%    passed     28.81%\n",
      "   7  AY675354.1     2.88%    passed     11.45%\n",
      "   8  DQ268652.1    93.88%    failed      2.20%\n",
      "   9  EF050454.1    93.88%    passed      9.06%\n",
      "  10  EF641806.1     2.33%    failed      1.82%\n",
      "  11  EF641807.1     2.33%    failed      2.42%\n",
      "  12  EU001330.1     2.26%    passed     35.14%\n",
      "  13  EU424341.1    94.98%    passed      8.00%\n",
      "  14  EU788002.1     2.44%    passed     91.64%\n",
      "  15  EU929078.1     8.28%    passed     22.53%\n",
      "  16  FJ593498.1     2.44%    failed      0.06%\n",
      "  17  FJ858378.1     2.48%    passed     14.89%\n",
      "  18  GQ200821.1     2.21%    passed     34.84%\n",
      "  19  GU566021.1    93.31%    passed      5.58%\n",
      "  20  GU904045.1     4.03%    failed      0.18%\n",
      "  21  GU997097.1     2.51%    passed     73.47%\n",
      "  22  HM015222.1     2.51%    failed      0.16%\n",
      "  23  HQ728461.1    93.71%    passed     18.21%\n",
      "  24  HQ834697.1     2.07%    passed     26.54%\n",
      "  25  JN037851.1    93.85%    passed      9.46%\n",
      "  26  JQ287716.2    74.86%    passed     98.19%\n",
      "  27  JX028271.1     2.72%    passed      9.90%\n",
      "  28  JX193697.1     2.50%    failed      0.12%\n",
      "  29  JX193700.1     1.84%    failed      0.00%\n",
      "  30  JX465369.1    94.90%    passed     56.89%\n",
      "  31  JX465379.1    95.18%    failed      0.27%\n",
      "  32  JX465381.1    95.18%    failed      0.05%\n",
      "  33  JX465382.1    95.18%    failed      0.11%\n",
      "  34  JX465383.1    95.18%    failed      0.12%\n",
      "  35  JX912953.2     3.08%    passed     49.50%\n",
      "  36  JX990941.1    88.38%    passed     85.68%\n",
      "  37  JX990946.1    84.75%    passed     43.98%\n",
      "  38  JX990965.1    88.38%    passed     95.90%\n",
      "  39  KC490923.1     3.43%    passed     32.55%\n",
      "  40  KC631784.1     2.62%    passed     40.61%\n",
      "  41  KC880348.1    74.30%    passed     29.95%\n",
      "  42  KF010517.1    88.09%    passed     74.97%\n",
      "  43  KF031355.1    88.09%    passed     85.44%\n",
      "  44  KF704713.1    91.57%    passed     55.78%\n",
      "  45  KF704715.1    88.05%    passed     85.10%\n",
      "  46  KF704716.1    82.76%    passed     90.18%\n",
      "  47  KF704717.1    82.76%    passed     82.14%\n",
      "  48  KF705679.1     2.87%    passed     95.93%\n",
      "  49  KF974361.1    36.58%    passed     53.80%\n",
      "  50  KJ000540.1    94.47%    passed     36.81%\n",
      "  51  KJ857315.1     2.20%    passed     71.82%\n",
      "  52  KJ857316.1     2.63%    passed     61.69%\n",
      "  53  KJ857320.1     3.39%    passed     88.98%\n",
      "  54  KM102249.1     2.91%    passed     53.18%\n",
      "  55  KR072623.1     2.44%    failed      1.74%\n",
      "  56  KR920360.1    94.51%    passed      8.20%\n",
      "  57  KT316176.1    46.75%    failed      0.29%\n",
      "  58  KT899703.1     2.91%    passed     19.49%\n",
      "  59  KU297981.1    94.84%    passed     33.24%\n",
      "  60  KU950715.1     2.54%    failed      0.56%\n",
      "  61  KX512432.1     2.44%    failed      0.70%\n",
      "  62  KX779125.1     3.06%    passed      5.38%\n",
      "  63  KX845680.1     4.39%    failed      0.00%\n",
      "  64  KY040508.1    89.70%    failed      3.18%\n",
      "  65  KY751731.1    94.35%    passed     71.24%\n",
      "  66  LC553715.1     2.94%    passed     22.48%\n",
      "  67  MF683846.1     2.81%    failed      0.00%\n",
      "  68  MG386252.1    94.84%    passed     27.26%\n",
      "  69  MG663536.1     2.85%    failed      1.74%\n",
      "  70  MK165653.1     4.16%    failed      0.00%\n",
      "  71  MK542664.1     3.55%    passed     41.79%\n",
      "  72  MK605664.1    97.10%    passed     72.92%\n",
      "  73  MK605665.1    97.10%    passed     84.51%\n",
      "  74  MK883761.1     3.89%    passed     52.57%\n",
      "  75  MN006903.1    79.60%    passed     38.47%\n",
      "  76  MN337866.1    94.31%    passed     13.05%\n",
      "  77  MN608087.1     3.57%    failed      0.85%\n",
      "  78  MN639746.1     2.62%    passed     15.36%\n",
      "  79  MN850095.1     2.96%    passed     83.08%\n",
      "  80  MT441741.1     3.18%    passed     85.21%\n",
      "  81  MT993951.1    94.17%    passed     18.65%\n",
      "  82  MW349026.1    94.66%    passed     86.66%\n",
      "  83  MZ191082.1    94.60%    passed     85.06%\n",
      "  84  NC_034401.1    2.54%    failed      0.56%\n",
      "  85  NC_038299.1    2.57%    failed      1.54%\n",
      "  86  NC_038515.1    2.91%    passed     53.18%\n",
      "  87  NC_043175.1   34.84%    passed     52.03%\n",
      "  88  NC_043407.1   94.01%    passed     54.28%\n",
      "  89  NC_055632.1    4.16%    failed      0.00%\n",
      "  90  NC_078262.1    3.05%    passed      7.31%\n",
      "  91  NC_078485.1    2.91%    failed      1.11%\n",
      "  92  OM912840.1    94.57%    failed      2.80%\n",
      "  93  OM912841.1    94.57%    failed      2.16%\n",
      "  94  OM912842.1    94.63%    failed      3.17%\n",
      "  95  OM912843.1    94.57%    failed      0.32%\n",
      "  96  OM912844.1    94.57%    failed      0.47%\n",
      "  97  OM963009.1    94.57%    failed      0.47%\n",
      "  98  ON571586.1     2.99%    passed     21.70%\n",
      "  99  ON571589.1     2.78%    passed     18.88%\n",
      " 100  OR047284.1    94.41%    passed     18.12%\n",
      " 101  OR684449.1     6.48%    failed      0.89%\n",
      " 102  OR934973.1    94.75%    failed      0.65%\n",
      " 103  OR934974.1    94.75%    failed      0.95%\n",
      "WARNING: 47 sequences contain more than 50% gaps/ambiguity\n",
      "****  TOTAL         44.66%  35 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: NC_038515.1 is identical to KM102249.1 but kept for subsequent analysis\n",
      "NOTE: NC_034401.1 is identical to KU950715.1 but kept for subsequent analysis\n",
      "NOTE: NC_055632.1 is identical to MK165653.1 but kept for subsequent analysis\n",
      "NOTE: OM963009.1 is identical to OM912844.1 but kept for subsequent analysis\n",
      "\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.120 seconds\n",
      "Measuring multi-threading efficiency up to 24 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "18 trees examined\n",
      "Threads: 1 / Time: 25.231 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -281729\n",
      "Threads: 2 / Time: 14.812 sec / Speedup: 1.703 / Efficiency: 85% / LogL: -281729\n",
      "Threads: 3 / Time: 10.479 sec / Speedup: 2.408 / Efficiency: 80% / LogL: -281729\n",
      "Threads: 4 / Time: 9.354 sec / Speedup: 2.697 / Efficiency: 67% / LogL: -281729\n",
      "Threads: 5 / Time: 8.590 sec / Speedup: 2.937 / Efficiency: 59% / LogL: -281729\n",
      "Threads: 6 / Time: 7.829 sec / Speedup: 3.223 / Efficiency: 54% / LogL: -281729\n",
      "Threads: 7 / Time: 7.121 sec / Speedup: 3.543 / Efficiency: 51% / LogL: -281729\n",
      "Threads: 8 / Time: 6.997 sec / Speedup: 3.606 / Efficiency: 45% / LogL: -281729\n",
      "BEST NUMBER OF THREADS: 7\n",
      "\n",
      "Perform fast likelihood tree search using GTR+I+G model...\n",
      "Estimate model parameters (epsilon = 5.000)\n",
      "Perform nearest neighbor interchange...\n",
      "Estimate model parameters (epsilon = 1.000)\n",
      "1. Initial log-likelihood: -209290.841\n",
      "2. Current log-likelihood: -209289.246\n",
      "Optimal log-likelihood: -209288.879\n",
      "Rate parameters:  A-C: 7.54670  A-G: 10.33749  A-T: 2.26263  C-G: 3.78682  C-T: 24.51188  G-T: 1.00000\n",
      "Base frequencies:  A: 0.333  C: 0.153  G: 0.210  T: 0.304\n",
      "Proportion of invariable sites: 0.233\n",
      "Gamma shape alpha: 0.639\n",
      "Parameters optimization took 2 rounds (0.563 sec)\n",
      "Time for fast ML tree search: 6.115 seconds\n",
      "\n",
      "NOTE: ModelFinder requires 167 MB RAM!\n",
      "ModelFinder will test up to 286 DNA models (sample size: 6727) ...\n",
      " No. Model         -LnL         df  AIC          AICc         BIC\n",
      "  1  GTR+F         241768.777   211 483959.554   483973.286   485397.284\n",
      "  2  GTR+F+I       224461.339   212 449346.677   449360.541   450791.221\n",
      "  3  GTR+F+G4      209777.787   212 419979.573   419993.438   421424.117\n",
      "  4  GTR+F+I+G4    209288.601   213 419003.202   419017.199   420454.559\n",
      "  5  GTR+F+R2      216405.979   213 433237.958   433251.955   434689.315\n",
      "  6  GTR+F+R3      212162.830   215 424755.661   424769.926   426220.646\n",
      "  7  GTR+F+R4      210911.826   217 422257.652   422272.187   423736.265\n",
      "  8  GTR+F+R5      210300.929   219 421039.857   421054.666   422532.098\n",
      "  9  GTR+F+R6      209895.530   221 420233.060   420248.144   421738.928\n",
      " 10  GTR+F+R7      209622.768   223 419691.537   419706.900   421211.033\n",
      " 11  GTR+F+R8      209440.415   225 419330.831   419346.475   420863.955\n",
      " 12  GTR+F+R9      209314.589   227 419083.179   419099.106   420629.931\n",
      " 13  GTR+F+R10     209267.335   229 418992.670   419008.884   420553.050\n",
      " 17  SYM+I+G4      213226.732   210 426873.463   426887.064   428304.379\n",
      " 30  TVM+F+I+G4    209998.757   212 420421.514   420435.379   421866.058\n",
      " 43  TVMe+I+G4     213281.895   209 426981.789   426995.258   428405.891\n",
      " 56  TIM3+F+I+G4   209646.315   211 419714.631   419728.363   421152.360\n",
      " 69  TIM3e+I+G4    214832.923   208 430081.847   430095.186   431499.135\n",
      " 82  TIM2+F+I+G4   210026.424   211 420474.848   420488.580   421912.577\n",
      " 95  TIM2e+I+G4    213663.830   208 427743.660   427756.999   429160.948\n",
      "108  TIM+F+I+G4    210406.863   211 421235.725   421249.457   422673.455\n",
      "121  TIMe+I+G4     214363.014   208 429142.027   429155.366   430559.315\n",
      "134  TPM3u+F+I+G4  210242.041   210 420904.082   420917.683   422334.998\n",
      "147  TPM3+F+I+G4   210242.041   210 420904.082   420917.683   422334.998\n",
      "160  TPM2u+F+I+G4  210554.489   210 421528.978   421542.579   422959.894\n",
      "173  TPM2+F+I+G4   210554.489   210 421528.978   421542.579   422959.894\n",
      "186  K3Pu+F+I+G4   210826.113   210 422072.226   422085.827   423503.142\n",
      "199  K3P+I+G4      214459.589   207 429333.178   429346.387   430743.652\n",
      "212  TN+F+I+G4     210411.138   210 421242.277   421255.877   422673.192\n",
      "225  TNe+I+G4      214951.885   207 430317.771   430330.980   431728.245\n",
      "238  HKY+F+I+G4    210841.155   209 422100.311   422113.780   423524.412\n",
      "251  K2P+I+G4      215057.173   206 430526.346   430539.426   431930.006\n",
      "264  F81+F+I+G4    218276.618   208 436969.236   436982.575   438386.524\n",
      "277  JC+I+G4       219643.023   205 439696.045   439708.997   441092.891\n",
      "Akaike Information Criterion:           GTR+F+R10\n",
      "Corrected Akaike Information Criterion: GTR+F+R10\n",
      "Bayesian Information Criterion:         GTR+F+I+G4\n",
      "Best-fit model: GTR+F+I+G4 chosen according to BIC\n",
      "\n",
      "All model information printed to tree_MF2.model.gz\n",
      "CPU time for ModelFinder: 1778.807 seconds (0h:29m:38s)\n",
      "Wall-clock time for ModelFinder: 301.512 seconds (0h:5m:1s)\n",
      "\n",
      "NOTE: 67 MB RAM (0 GB) is required!\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "Thoroughly optimizing +I+G parameters from 10 start values...\n",
      "Init pinv, alpha: 0.000, 0.638 / Estimate: 0.000, 0.385 / LogL: -209777.808\n",
      "Init pinv, alpha: 0.030, 0.638 / Estimate: 0.234, 0.638 / LogL: -209288.699\n",
      "Init pinv, alpha: 0.061, 0.638 / Estimate: 0.234, 0.637 / LogL: -209288.569\n",
      "Init pinv, alpha: 0.091, 0.638 / Estimate: 0.234, 0.637 / LogL: -209288.554\n",
      "Init pinv, alpha: 0.121, 0.638 / Estimate: 0.234, 0.637 / LogL: -209288.566\n",
      "Init pinv, alpha: 0.152, 0.638 / Estimate: 0.234, 0.638 / LogL: -209288.565\n",
      "Init pinv, alpha: 0.182, 0.638 / Estimate: 0.234, 0.637 / LogL: -209288.544\n",
      "Init pinv, alpha: 0.213, 0.638 / Estimate: 0.233, 0.637 / LogL: -209288.531\n",
      "Init pinv, alpha: 0.243, 0.638 / Estimate: 0.234, 0.637 / LogL: -209288.489\n",
      "Init pinv, alpha: 0.273, 0.638 / Estimate: 0.234, 0.637 / LogL: -209288.507\n",
      "Optimal pinv,alpha: 0.234, 0.637 / LogL: -209288.489\n",
      "\n",
      "Parameters optimization took 19.243 sec\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.045802 sec (of wall-clock time) 0.290741 sec(of CPU time)\n",
      "WARNING: Some pairwise ML distances are too long (saturated)\n",
      "Computing RapidNJ tree took 0.192887 sec (of wall-clock time) 0.241071 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -210632.021\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 13.941 second\n",
      "Computing log-likelihood of 98 initial trees ... 5.343 seconds\n",
      "Current best score: -209288.489\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -209281.158\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 10: -209279.391\n",
      "Iteration 10 / LogL: -209279.391 / Time: 0h:0m:53s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 16: -209268.953\n",
      "Iteration 20 / LogL: -209285.970 / Time: 0h:1m:6s\n",
      "Finish initializing candidate tree set (12)\n",
      "Current best tree score: -209268.953 / CPU time: 45.561\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -209292.453 / Time: 0h:1m:15s (0h:3m:44s left)\n",
      "Iteration 40 / LogL: -209298.967 / Time: 0h:1m:24s (0h:2m:45s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 49: -209267.797\n",
      "Iteration 50 / LogL: -209276.663 / Time: 0h:1m:35s (0h:3m:12s left)\n",
      "Iteration 60 / LogL: -209268.350 / Time: 0h:1m:47s (0h:2m:41s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -209267.743\n",
      "Iteration 70 / LogL: -209275.827 / Time: 0h:1m:59s (0h:2m:16s left)\n",
      "Iteration 80 / LogL: -209278.484 / Time: 0h:2m:11s (0h:1m:54s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -209267.653\n",
      "Iteration 90 / LogL: -209267.653 / Time: 0h:2m:23s (0h:1m:34s left)\n",
      "Iteration 100 / LogL: -209275.271 / Time: 0h:2m:35s (0h:1m:16s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 105: -209267.371\n",
      "Iteration 110 / LogL: -209271.915 / Time: 0h:2m:49s (0h:2m:27s left)\n",
      "Iteration 120 / LogL: -209341.553 / Time: 0h:3m:2s (0h:2m:10s left)\n",
      "Iteration 130 / LogL: -209267.745 / Time: 0h:3m:18s (0h:1m:55s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -209267.266\n",
      "Iteration 140 / LogL: -209268.442 / Time: 0h:3m:31s (0h:1m:39s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -209267.248\n",
      "Iteration 150 / LogL: -209283.064 / Time: 0h:3m:45s (0h:1m:23s left)\n",
      "Iteration 160 / LogL: -209269.183 / Time: 0h:3m:58s (0h:1m:7s left)\n",
      "Iteration 170 / LogL: -209268.974 / Time: 0h:4m:10s (0h:0m:51s left)\n",
      "Iteration 180 / LogL: -209273.443 / Time: 0h:4m:25s (0h:0m:37s left)\n",
      "Iteration 190 / LogL: -209270.022 / Time: 0h:4m:37s (0h:0m:22s left)\n",
      "Iteration 200 / LogL: -209292.210 / Time: 0h:4m:49s (0h:0m:7s left)\n",
      "TREE SEARCH COMPLETED AFTER 206 ITERATIONS / Time: 0h:5m:0s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -209267.248\n",
      "Optimal log-likelihood: -209267.247\n",
      "Rate parameters:  A-C: 7.80299  A-G: 10.65398  A-T: 2.31085  C-G: 3.90797  C-T: 25.48194  G-T: 1.00000\n",
      "Base frequencies:  A: 0.333  C: 0.153  G: 0.210  T: 0.304\n",
      "Proportion of invariable sites: 0.233\n",
      "Gamma shape alpha: 0.635\n",
      "Parameters optimization took 1 rounds (0.209 sec)\n",
      "BEST SCORE FOUND : -209267.247\n",
      "Total tree length: 34.105\n",
      "\n",
      "Total number of iterations: 206\n",
      "CPU time used for tree search: 1867.176 sec (0h:31m:7s)\n",
      "Wall-clock time used for tree search: 279.669 sec (0h:4m:39s)\n",
      "Total CPU time used: 2009.972 sec (0h:33m:29s)\n",
      "Total wall-clock time used: 300.524 sec (0h:5m:0s)\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                tree_MF2.iqtree\n",
      "  Maximum-likelihood tree:       tree_MF2.treefile\n",
      "  Likelihood distances:          tree_MF2.mldist\n",
      "  Screen log file:               tree_MF2.log\n",
      "\n",
      "Date and Time: Sat Apr 20 14:46:58 2024\n"
     ]
    }
   ],
   "source": [
    "! iqtree2 -m MFP -s data/all_seqs_mafft.fa --prefix model-finder/tree_MF2 -T AUTO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best-fit model according to BIC: GTR+F+I+G4\n",
      "\n",
      "List of models sorted by BIC scores: \n",
      "\n",
      "Model                  LogL         AIC      w-AIC        AICc     w-AICc         BIC      w-BIC\n",
      "GTR+F+I+G4      -209288.601  419003.202 -  0.00514  419017.199 -   0.0154  420454.559 +        1\n"
     ]
    }
   ],
   "source": [
    "! head -42 model-finder/tree_MF2.iqtree | tail -6"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 5. Building the final tree**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`iq-tree2` launch with the best substitution model & generating 1000 replicas of ultrafast bootstrap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IQ-TREE multicore version 2.1.4-beta COVID-edition for Linux 64-bit built Jun 24 2021\n",
      "Developed by Bui Quang Minh, James Barbetti, Nguyen Lam Tung,\n",
      "Olga Chernomor, Heiko Schmidt, Dominik Schrempf, Michael Woodhams.\n",
      "\n",
      "Host:    IliaPC (AVX2, FMA3, 15 GB RAM)\n",
      "Command: iqtree2 -s all_seqs_mafft.fa -m GTR+F+I+G4 -pre tree_ufb -bb 1000 -nt AUTO\n",
      "Seed:    782323 (Using SPRNG - Scalable Parallel Random Number Generator)\n",
      "Time:    Sat Apr 20 14:48:04 2024\n",
      "Kernel:  AVX+FMA - auto-detect threads (24 CPU cores detected)\n",
      "\n",
      "Reading alignment file all_seqs_mafft.fa ... Fasta format detected\n",
      "Alignment most likely contains DNA/RNA sequences\n",
      "Alignment has 103 sequences with 6727 columns, 5018 distinct patterns\n",
      "4543 parsimony-informative, 346 singleton sites, 1838 constant sites\n",
      "             Gap/Ambiguity  Composition  p-value\n",
      "   1  AB620030.1     3.46%    passed     70.90%\n",
      "   2  AB620102.1     2.94%    passed      9.29%\n",
      "   3  AB620105.1     3.12%    passed     42.74%\n",
      "   4  AB677488.1     5.11%    passed     11.73%\n",
      "   5  AF005729.1    92.26%    passed     59.80%\n",
      "   6  AM998806.1    93.88%    passed     28.81%\n",
      "   7  AY675354.1     2.88%    passed     11.45%\n",
      "   8  DQ268652.1    93.88%    failed      2.20%\n",
      "   9  EF050454.1    93.88%    passed      9.06%\n",
      "  10  EF641806.1     2.33%    failed      1.82%\n",
      "  11  EF641807.1     2.33%    failed      2.42%\n",
      "  12  EU001330.1     2.26%    passed     35.14%\n",
      "  13  EU424341.1    94.98%    passed      8.00%\n",
      "  14  EU788002.1     2.44%    passed     91.64%\n",
      "  15  EU929078.1     8.28%    passed     22.53%\n",
      "  16  FJ593498.1     2.44%    failed      0.06%\n",
      "  17  FJ858378.1     2.48%    passed     14.89%\n",
      "  18  GQ200821.1     2.21%    passed     34.84%\n",
      "  19  GU566021.1    93.31%    passed      5.58%\n",
      "  20  GU904045.1     4.03%    failed      0.18%\n",
      "  21  GU997097.1     2.51%    passed     73.47%\n",
      "  22  HM015222.1     2.51%    failed      0.16%\n",
      "  23  HQ728461.1    93.71%    passed     18.21%\n",
      "  24  HQ834697.1     2.07%    passed     26.54%\n",
      "  25  JN037851.1    93.85%    passed      9.46%\n",
      "  26  JQ287716.2    74.86%    passed     98.19%\n",
      "  27  JX028271.1     2.72%    passed      9.90%\n",
      "  28  JX193697.1     2.50%    failed      0.12%\n",
      "  29  JX193700.1     1.84%    failed      0.00%\n",
      "  30  JX465369.1    94.90%    passed     56.89%\n",
      "  31  JX465379.1    95.18%    failed      0.27%\n",
      "  32  JX465381.1    95.18%    failed      0.05%\n",
      "  33  JX465382.1    95.18%    failed      0.11%\n",
      "  34  JX465383.1    95.18%    failed      0.12%\n",
      "  35  JX912953.2     3.08%    passed     49.50%\n",
      "  36  JX990941.1    88.38%    passed     85.68%\n",
      "  37  JX990946.1    84.75%    passed     43.98%\n",
      "  38  JX990965.1    88.38%    passed     95.90%\n",
      "  39  KC490923.1     3.43%    passed     32.55%\n",
      "  40  KC631784.1     2.62%    passed     40.61%\n",
      "  41  KC880348.1    74.30%    passed     29.95%\n",
      "  42  KF010517.1    88.09%    passed     74.97%\n",
      "  43  KF031355.1    88.09%    passed     85.44%\n",
      "  44  KF704713.1    91.57%    passed     55.78%\n",
      "  45  KF704715.1    88.05%    passed     85.10%\n",
      "  46  KF704716.1    82.76%    passed     90.18%\n",
      "  47  KF704717.1    82.76%    passed     82.14%\n",
      "  48  KF705679.1     2.87%    passed     95.93%\n",
      "  49  KF974361.1    36.58%    passed     53.80%\n",
      "  50  KJ000540.1    94.47%    passed     36.81%\n",
      "  51  KJ857315.1     2.20%    passed     71.82%\n",
      "  52  KJ857316.1     2.63%    passed     61.69%\n",
      "  53  KJ857320.1     3.39%    passed     88.98%\n",
      "  54  KM102249.1     2.91%    passed     53.18%\n",
      "  55  KR072623.1     2.44%    failed      1.74%\n",
      "  56  KR920360.1    94.51%    passed      8.20%\n",
      "  57  KT316176.1    46.75%    failed      0.29%\n",
      "  58  KT899703.1     2.91%    passed     19.49%\n",
      "  59  KU297981.1    94.84%    passed     33.24%\n",
      "  60  KU950715.1     2.54%    failed      0.56%\n",
      "  61  KX512432.1     2.44%    failed      0.70%\n",
      "  62  KX779125.1     3.06%    passed      5.38%\n",
      "  63  KX845680.1     4.39%    failed      0.00%\n",
      "  64  KY040508.1    89.70%    failed      3.18%\n",
      "  65  KY751731.1    94.35%    passed     71.24%\n",
      "  66  LC553715.1     2.94%    passed     22.48%\n",
      "  67  MF683846.1     2.81%    failed      0.00%\n",
      "  68  MG386252.1    94.84%    passed     27.26%\n",
      "  69  MG663536.1     2.85%    failed      1.74%\n",
      "  70  MK165653.1     4.16%    failed      0.00%\n",
      "  71  MK542664.1     3.55%    passed     41.79%\n",
      "  72  MK605664.1    97.10%    passed     72.92%\n",
      "  73  MK605665.1    97.10%    passed     84.51%\n",
      "  74  MK883761.1     3.89%    passed     52.57%\n",
      "  75  MN006903.1    79.60%    passed     38.47%\n",
      "  76  MN337866.1    94.31%    passed     13.05%\n",
      "  77  MN608087.1     3.57%    failed      0.85%\n",
      "  78  MN639746.1     2.62%    passed     15.36%\n",
      "  79  MN850095.1     2.96%    passed     83.08%\n",
      "  80  MT441741.1     3.18%    passed     85.21%\n",
      "  81  MT993951.1    94.17%    passed     18.65%\n",
      "  82  MW349026.1    94.66%    passed     86.66%\n",
      "  83  MZ191082.1    94.60%    passed     85.06%\n",
      "  84  NC_034401.1    2.54%    failed      0.56%\n",
      "  85  NC_038299.1    2.57%    failed      1.54%\n",
      "  86  NC_038515.1    2.91%    passed     53.18%\n",
      "  87  NC_043175.1   34.84%    passed     52.03%\n",
      "  88  NC_043407.1   94.01%    passed     54.28%\n",
      "  89  NC_055632.1    4.16%    failed      0.00%\n",
      "  90  NC_078262.1    3.05%    passed      7.31%\n",
      "  91  NC_078485.1    2.91%    failed      1.11%\n",
      "  92  OM912840.1    94.57%    failed      2.80%\n",
      "  93  OM912841.1    94.57%    failed      2.16%\n",
      "  94  OM912842.1    94.63%    failed      3.17%\n",
      "  95  OM912843.1    94.57%    failed      0.32%\n",
      "  96  OM912844.1    94.57%    failed      0.47%\n",
      "  97  OM963009.1    94.57%    failed      0.47%\n",
      "  98  ON571586.1     2.99%    passed     21.70%\n",
      "  99  ON571589.1     2.78%    passed     18.88%\n",
      " 100  OR047284.1    94.41%    passed     18.12%\n",
      " 101  OR684449.1     6.48%    failed      0.89%\n",
      " 102  OR934973.1    94.75%    failed      0.65%\n",
      " 103  OR934974.1    94.75%    failed      0.95%\n",
      "WARNING: 47 sequences contain more than 50% gaps/ambiguity\n",
      "****  TOTAL         44.66%  35 sequences failed composition chi2 test (p-value<5%; df=3)\n",
      "NOTE: NC_038515.1 is identical to KM102249.1 but kept for subsequent analysis\n",
      "NOTE: NC_034401.1 is identical to KU950715.1 but kept for subsequent analysis\n",
      "NOTE: NC_055632.1 is identical to MK165653.1 but kept for subsequent analysis\n",
      "NOTE: OM963009.1 is identical to OM912844.1 but kept for subsequent analysis\n",
      "\n",
      "Create initial parsimony tree by phylogenetic likelihood library (PLL)... 0.125 seconds\n",
      "Generating 1000 samples for ultrafast bootstrap (seed: 782323)...\n",
      "\n",
      "NOTE: 86 MB RAM (0 GB) is required!\n",
      "Measuring multi-threading efficiency up to 24 CPU cores\n",
      "Increase to 10 rounds for branch lengths\n",
      "18 trees examined\n",
      "Threads: 1 / Time: 24.980 sec / Speedup: 1.000 / Efficiency: 100% / LogL: -279813\n",
      "Threads: 2 / Time: 15.046 sec / Speedup: 1.660 / Efficiency: 83% / LogL: -279813\n",
      "Threads: 3 / Time: 11.519 sec / Speedup: 2.169 / Efficiency: 72% / LogL: -279813\n",
      "Threads: 4 / Time: 10.426 sec / Speedup: 2.396 / Efficiency: 60% / LogL: -279813\n",
      "Threads: 5 / Time: 7.656 sec / Speedup: 3.263 / Efficiency: 65% / LogL: -279813\n",
      "Threads: 6 / Time: 7.392 sec / Speedup: 3.379 / Efficiency: 56% / LogL: -279813\n",
      "Threads: 7 / Time: 6.835 sec / Speedup: 3.655 / Efficiency: 52% / LogL: -279813\n",
      "Threads: 8 / Time: 6.128 sec / Speedup: 4.076 / Efficiency: 51% / LogL: -279813\n",
      "Threads: 9 / Time: 6.252 sec / Speedup: 3.996 / Efficiency: 44% / LogL: -279813\n",
      "BEST NUMBER OF THREADS: 8\n",
      "\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "Thoroughly optimizing +I+G parameters from 10 start values...\n",
      "Init pinv, alpha: 0.000, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.016\n",
      "Init pinv, alpha: 0.030, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.264\n",
      "Init pinv, alpha: 0.061, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.463\n",
      "Init pinv, alpha: 0.091, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.130\n",
      "Init pinv, alpha: 0.121, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.430\n",
      "Init pinv, alpha: 0.152, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.700\n",
      "Init pinv, alpha: 0.182, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.883\n",
      "Init pinv, alpha: 0.213, 1.000 / Estimate: 0.235, 0.644 / LogL: -209621.008\n",
      "Init pinv, alpha: 0.243, 1.000 / Estimate: 0.235, 0.644 / LogL: -209621.053\n",
      "Init pinv, alpha: 0.273, 1.000 / Estimate: 0.235, 0.645 / LogL: -209621.815\n",
      "Optimal pinv,alpha: 0.235, 0.644 / LogL: -209621.008\n",
      "\n",
      "Parameters optimization took 40.204 sec\n",
      "Computing ML distances based on estimated model parameters...\n",
      "Computing ML distances took 0.061960 sec (of wall-clock time) 0.354195 sec(of CPU time)\n",
      "WARNING: Some pairwise ML distances are too long (saturated)\n",
      "Computing RapidNJ tree took 0.232801 sec (of wall-clock time) 0.268872 sec (of CPU time)\n",
      "Log-likelihood of RapidNJ tree: -210868.113\n",
      "--------------------------------------------------------------------\n",
      "|             INITIALIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Generating 98 parsimony trees... 14.179 second\n",
      "Computing log-likelihood of 98 initial trees ... 4.775 seconds\n",
      "Current best score: -209621.008\n",
      "\n",
      "Do NNI search on 20 best initial trees\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 1: -209388.659\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 2: -209293.650\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 4: -209279.436\n",
      "Iteration 10 / LogL: -209279.920 / Time: 0h:2m:55s\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 14: -209268.884\n",
      "Iteration 20 / LogL: -209273.222 / Time: 0h:3m:16s\n",
      "Finish initializing candidate tree set (12)\n",
      "Current best tree score: -209268.884 / CPU time: 58.888\n",
      "Number of iterations: 20\n",
      "--------------------------------------------------------------------\n",
      "|               OPTIMIZING CANDIDATE TREE SET                      |\n",
      "--------------------------------------------------------------------\n",
      "Iteration 30 / LogL: -209281.874 / Time: 0h:3m:31s (0h:10m:13s left)\n",
      "Iteration 40 / LogL: -209275.757 / Time: 0h:3m:45s (0h:7m:7s left)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -209268.880\n",
      "Iteration 50 / LogL: -209278.675 / Time: 0h:4m:2s (0h:5m:16s left)\n",
      "Log-likelihood cutoff on original alignment: -209338.679\n",
      "Iteration 60 / LogL: -209270.378 / Time: 0h:4m:19s (0h:3m:57s left)\n",
      "Iteration 70 / LogL: -209276.662 / Time: 0h:4m:37s (0h:2m:57s left)\n",
      "Iteration 80 / LogL: -209269.863 / Time: 0h:4m:51s (0h:2m:5s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 89: -209268.588\n",
      "Iteration 90 / LogL: -209278.529 / Time: 0h:5m:5s (0h:5m:39s left)\n",
      "Iteration 100 / LogL: -209462.045 / Time: 0h:5m:15s (0h:4m:43s left)\n",
      "Log-likelihood cutoff on original alignment: -209338.679\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.996\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "UPDATE BEST LOG-LIKELIHOOD: -209268.429\n",
      "Iteration 110 / LogL: -209285.950 / Time: 0h:5m:29s (0h:4m:32s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 113: -209268.048\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 118: -209267.483\n",
      "Iteration 120 / LogL: -209268.889 / Time: 0h:5m:44s (0h:4m:43s left)\n",
      "Iteration 130 / LogL: -209271.218 / Time: 0h:6m:0s (0h:4m:5s left)\n",
      "Estimate model parameters (epsilon = 0.100)\n",
      "BETTER TREE FOUND at iteration 134: -209267.246\n",
      "Iteration 140 / LogL: -209324.717 / Time: 0h:6m:13s (0h:4m:12s left)\n",
      "Iteration 150 / LogL: -209270.327 / Time: 0h:6m:27s (0h:3m:38s left)\n",
      "Log-likelihood cutoff on original alignment: -209333.756\n",
      "Iteration 160 / LogL: -209271.106 / Time: 0h:6m:42s (0h:3m:7s left)\n",
      "Iteration 170 / LogL: -209267.694 / Time: 0h:6m:57s (0h:2m:37s left)\n",
      "Iteration 180 / LogL: -209269.642 / Time: 0h:7m:8s (0h:2m:9s left)\n",
      "Iteration 190 / LogL: -209272.533 / Time: 0h:7m:25s (0h:1m:43s left)\n",
      "Iteration 200 / LogL: -209268.892 / Time: 0h:7m:38s (0h:1m:18s left)\n",
      "Log-likelihood cutoff on original alignment: -209333.756\n",
      "NOTE: Bootstrap correlation coefficient of split occurrence frequencies: 0.996\n",
      "Iteration 210 / LogL: -209267.360 / Time: 0h:7m:52s (0h:3m:23s left)\n",
      "Iteration 220 / LogL: -209268.945 / Time: 0h:8m:4s (0h:2m:57s left)\n",
      "Iteration 230 / LogL: -209270.478 / Time: 0h:8m:16s (0h:2m:31s left)\n",
      "TREE SEARCH COMPLETED AFTER 235 ITERATIONS / Time: 0h:8m:24s\n",
      "\n",
      "--------------------------------------------------------------------\n",
      "|                    FINALIZING TREE SEARCH                        |\n",
      "--------------------------------------------------------------------\n",
      "Performs final model parameters optimization\n",
      "Estimate model parameters (epsilon = 0.010)\n",
      "1. Initial log-likelihood: -209267.246\n",
      "Optimal log-likelihood: -209267.244\n",
      "Rate parameters:  A-C: 7.77919  A-G: 10.61797  A-T: 2.30730  C-G: 3.89680  C-T: 25.41334  G-T: 1.00000\n",
      "Base frequencies:  A: 0.333  C: 0.153  G: 0.210  T: 0.304\n",
      "Proportion of invariable sites: 0.234\n",
      "Gamma shape alpha: 0.636\n",
      "Parameters optimization took 1 rounds (0.169 sec)\n",
      "BEST SCORE FOUND : -209267.244\n",
      "Creating bootstrap support values...\n",
      "Split supports printed to NEXUS file tree_ufb.splits.nex\n",
      "Total tree length: 34.094\n",
      "\n",
      "Total number of iterations: 235\n",
      "CPU time used for tree search: 2807.961 sec (0h:46m:47s)\n",
      "Wall-clock time used for tree search: 366.850 sec (0h:6m:6s)\n",
      "Total CPU time used: 3500.367 sec (0h:58m:20s)\n",
      "Total wall-clock time used: 505.440 sec (0h:8m:25s)\n",
      "\n",
      "Computing bootstrap consensus tree...\n",
      "Reading input file tree_ufb.splits.nex...\n",
      "103 taxa and 392 splits.\n",
      "Consensus tree written to tree_ufb.contree\n",
      "Reading input trees file tree_ufb.contree\n",
      "Log-likelihood of consensus tree: -209267.684\n",
      "\n",
      "Analysis results written to: \n",
      "  IQ-TREE report:                tree_ufb.iqtree\n",
      "  Maximum-likelihood tree:       tree_ufb.treefile\n",
      "  Likelihood distances:          tree_ufb.mldist\n",
      "\n",
      "Ultrafast bootstrap approximation results written to:\n",
      "  Split support values:          tree_ufb.splits.nex\n",
      "  Consensus tree:                tree_ufb.contree\n",
      "  Screen log file:               tree_ufb.log\n",
      "\n",
      "Date and Time: Sat Apr 20 14:56:31 2024\n"
     ]
    }
   ],
   "source": [
    "! iqtree2 -s data/all_seqs_mafft.fa -m GTR+F+I+G4 -pre tree/tree_ufb -bb 1000 -nt AUTO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 6. First tree visualisation**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this purpose [iTOL](https://itol.embl.de) software was used with followed tree annotation in Pixelmator Pro.<br>\n",
    "To visualise the tree `tree_ufb.treefile` must be uploaded to [iTOL](https://itol.embl.de)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 7. Annotating the tree**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 7.1. Returning organisms names to the tree**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The request has been fulfilled.\n",
      "File saved to data/accession_organism.txt\n"
     ]
    }
   ],
   "source": [
    "get_organisms('iljapopov17@gmail.com', 'data/accession_numbers.txt', 'data/accession_organism.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_038515.1 Laibin virus\n",
      "KT316176.1 Makokou virus\n",
      "MN337866.1 Sarawak mobatvirus\n",
      "MG663536.1 Dakrong virus\n",
      "JX193700.1 Kilimanjaro virus\n"
     ]
    }
   ],
   "source": [
    "! head -5 data/accession_organism.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Everything worked well."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 7.2. Updating the tree**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The request has been fulfilled.\n",
      "File saved to tree/annotated_tree.treefile\n"
     ]
    }
   ],
   "source": [
    "update_tree('data/accession_organism.txt', 'tree/tree_ufb.treefile', 'tree/annotated_tree.treefile')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(AB620030.1 Amur virus:0.0771041878,((((((((((AB620102.1 Orthohantavirus montanoense:0.3586812403,AB620105.1 Carrizal virus:0.3164538911)100:0.0821256184,(ON571586.1 Orthohantavirus sinnombreense:0.0003496417,ON571589.1 Orthohantavirus sinnombreense:0.0000027792)100:0.3526346556)94:0.0679804387,(((((AF005729.1 Orthohantavirus negraense:0.3758037094,NC_043407.1 Necocli virus:0.3953678541)99:0.0611167013,MN850095.1 Orthohantavirus andesense:0.2115210053)99:0.1127135879,EU788002.1 Maporal virus:0.3285749060)99:0.0815653694,GQ200821.1 Orthohantavirus delgaditoense:0.3673061736)78:0.0440712496,(FJ858378.1 Catacamas virus:0.2198999425,(GU997097.1 Orthohantavirus nigrorivense:0.2424566531,NC_038299.1 Orthohantavirus bayoui:0.2143819459)97:0.0646432754)100:0.1905032908)100:0.0596284326)100:0.1483520267,((((AB677488.1 Ussuri virus:0.2315952341,JX028271.1 Muju virus:0.2947312110)57:0.0621199827,MN639746.1 Orthohantavirus puumalaense:0.2298150170)100:0.1004645997,(((((HQ728461.1 Orthohantavirus tulaense:0.0991419758,MT993951.1 Orthohantavirus tulaense:0.1198827908)100:0.0800585338,KU297981.1 Orthohantavirus tulaense:0.1618395760)100:0.1912227809,(((KY751731.1 Kielder hantavirus:0.1151455066,MK883761.1 Orthohantavirus tatenalense:0.1215848998)100:0.0990470187,MK542664.1 Traemmersee orthohantavirus:0.1986814928)100:0.0574368767,MT441741.1 Rusne orthohantavirus:0.1899661246)100:0.0618239866)98:0.0857654171,KJ857315.1 Orthohantavirus khabarovskense:0.2791373880)84:0.0473592676,KJ857316.1 Yuanjiang virus:0.2726077642)95:0.0542967810)100:0.1992801697,KT899703.1 Fugong virus:0.4700677119)100:0.1130002341)69:0.0567435563,HM015222.1 Rockport virus:0.6082931559)100:0.2677252605,(((((EF641806.1 Imjin virus:0.0146151276,KX779125.1 Imjin virus:0.0138818167)100:0.0769890463,EF641807.1 Imjin virus:0.0804300980)100:0.3346745369,EU001330.1 Thottapalayam virus:0.4228360394)100:0.2536585750,(JX193697.1 Uluguru virus:0.6692053051,JX193700.1 Kilimanjaro virus:0.4766759736)100:0.1298513865)100:0.4854320839,((((((EU424341.1 Altai virus:0.2654389245,NC_078485.1 Lena virus:0.2155769500)100:0.5582020229,JX990965.1 Boginia virus:0.5363602544)78:0.3823058529,((((FJ593498.1 Nova virus:0.1138267013,(KF010517.1 Nova virus:0.1090607191,KF031355.1 Nova virus:0.1938245416)98:0.0302351889)98:0.0252765011,KX512432.1 Nova virus:0.1341231385)99:0.0561960783,(JX990946.1 Nova virus:0.0340618990,KR072623.1 Nova virus:0.0079225951)100:0.0732037168)100:0.2867605796,KY040508.1 Asturias virus:0.2039299345)100:0.5707414930)58:0.0916991172,(((JN037851.1 Hantavirus MGB/1209:0.6361097514,KT316176.1 Makokou virus:0.3812644990)66:0.0456683944,((JX912953.2 Xuan son virus:0.1408219222,KF704713.1 Xuan son virus:0.1620798600)100:0.1150648055,((KF704715.1 Xuan son virus:0.0038421865,(KF704716.1 Xuan son virus:0.0062914844,KF704717.1 Xuan son virus:0.0145854887)100:0.0188894736)100:0.1802940875,NC_078262.1 Xuan son virus:0.1752778191)98:0.0584955771)100:0.2695940521)90:0.0551921386,(((KM102249.1 Laibin virus:0.0000010001,NC_038515.1 Laibin virus:0.0000010001)100:0.2426973377,MN337866.1 Sarawak mobatvirus:0.2544616999)98:0.1417640648,MG663536.1 Dakrong virus:0.4933778150)99:0.0875153991)98:0.2337792501)57:0.0683692102,((JX465369.1 Huangpi virus:0.5393286223,OR684449.1 Buritiense virus:0.5548160827)47:0.1175896767,((KU950715.1 Quezon virus:0.0000010001,NC_034401.1 Quezon virus:0.0000010001)100:0.5732888012,(MK165653.1 Orthohantavirus robinaense:0.0000010001,NC_055632.1 Orthohantavirus robinaense:0.0000010001)100:0.5309016345)100:0.2607446493)41:0.0898181287)42:0.0764483214,((JQ287716.2 Mouyassue virus:0.0000023848,KJ000540.1 Mouyassue virus:0.0000010001)100:0.4267644543,((JX465379.1 Longquan virus:0.0000020853,((JX465381.1 Longquan virus:0.0038141945,JX465382.1 Longquan virus:0.0077268601)100:0.0196543002,JX465383.1 Longquan virus:0.0000020681)96:0.0077421990)100:0.3703425331,(KR920360.1 Brno virus:0.0000021791,((((KX845680.1 Brno virus:0.0037963581,(OM912843.1 Loanvirus brunaense:0.0137485885,(OM912844.1 Loanvirus brunaense:0.0000010001,OM963009.1 Loanvirus brunaense:0.0000010001)69:0.0000020133)97:0.0033926390)55:0.0000026778,(OM912840.1 Loanvirus brunaense:0.0000010001,OM912842.1 Loanvirus brunaense:0.0000010001)99:0.0134331563)100:0.0542071098,(OR934973.1 Loanvirus brunaense:0.0049500299,OR934974.1 Loanvirus brunaense:0.0163547086)95:0.0197518443)95:0.0236773267,OM912841.1 Loanvirus brunaense:0.0180655310)39:0.0207196720)100:0.4180488781)90:0.1910070183)51:0.2445995949)76:0.1260525283)79:0.2957849738)100:0.2132251977,((((EF050454.1 Tanganya virus:0.4626176976,KC631784.1 Bowe virus:0.3150541234)99:0.1191717914,HQ834697.1 Jeju virus:0.4504119611)100:0.1997444220,(((EU929078.1 Asama virus:0.3651385985,((GU566021.1 Qian Hu Shan virus:0.2054115794,KC880348.1 Orthohantavirus asikkalaense:0.2583087016)80:0.0478747851,((JX990941.1 Seewis virus:0.2461045089,KF974361.1 Amga virus:0.2324359653)99:0.0505468662,KJ857320.1 Kenkeme virus:0.2844332666)86:0.0578482552)94:0.0883369661)99:0.0719271114,(KF705679.1 Xinyi virus:0.2888953092,MN006903.1 Jemez Springs virus:0.3470836938)100:0.0839674435)100:0.0520037869,NC_043175.1 Oxbow virus:0.5409916277)100:0.0713003641)100:0.0952947788,MF683846.1 Orthohantavirus brugesense:0.6818134739)59:0.0500300676)100:0.2567808467,((AM998806.1 Serang virus:0.1469951860,(KC490923.1 Anjozorobe virus:0.2429580339,LC553715.1 Orthohantavirus thailandense:0.2435300852)83:0.0474757758)100:0.1384906244,((GU904045.1 Orthohantavirus dobravaense:0.0902845208,MK605664.1 Orthohantavirus dobravaense:0.1225113023)98:0.0606089420,MK605665.1 Orthohantavirus dobravaense:0.0399085736)100:0.2020387586)89:0.0690533160)85:0.0190746178,(DQ268652.1 Orthohantavirus sangassouense:0.3466753654,(MG386252.1 Orthohantavirus seoulense:0.0137372170,OR047284.1 Orthohantavirus seoulense:0.0548624754)100:0.2848089750)91:0.0911279902)100:0.1716524032,(MN608087.1 Orthohantavirus hantanense:0.1744392228,(MW349026.1 Orthohantavirus hantanense:0.0248825313,MZ191082.1 Orthohantavirus hantanense:0.0110645270)100:0.1996850841)99:0.0404345463)100:0.1049070173,AY675354.1 Soochong virus-2:0.0948964879);\n"
     ]
    }
   ],
   "source": [
    "! head -1 tree/annotated_tree.treefile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now tree leaves have accession number + organism name (AB620030.1 Amur virus) instead of just an accession number (AB620030.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 7.3. Fetching information about viruses hosts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The request has been fulfilled.\n",
      "File saved to data/accession_host.txt\n"
     ]
    }
   ],
   "source": [
    "get_hosts('iljapopov17@gmail.com', 'data/accession_numbers.txt', 'data/accession_host.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_038515.1 Taphozous melanopogon\n",
      "KT316176.1 Hipposideros ruber\n",
      "MN337866.1 Murina aenea\n",
      "MG663536.1 Aselliscus stoliczkanus (Stoliczka's Asian trident bat)\n",
      "JX193700.1 Myosorex zinki\n"
     ]
    }
   ],
   "source": [
    "! head -5 data/accession_host.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Everything worked well."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 7.4. Fetching information about viruses hosts's phylogenetic order**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The request has been fulfilled.\n",
      "File saved to data/accession_order.txt\n",
      "Please do not forget to edit the file manually.\n",
      "The query to NCBI database from this function is pretty difficult.\n",
      "Sometimes this function prints:\n",
      "\"Error - HTTP Error 400: Bad Request\" in case of bad connection or\n",
      "\"Note - False record\" in case there is no record about the host organism.\n"
     ]
    }
   ],
   "source": [
    "get_hosts_orders('iljapopov17@gmail.com', 'data/accession_host.txt', 'data/accession_order.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_038515.1\tChiroptera\n",
      "KT316176.1\tChiroptera\n",
      "MN337866.1\tChiroptera\n",
      "MG663536.1\tChiroptera\n",
      "JX193700.1\tEulipotyphla\n"
     ]
    }
   ],
   "source": [
    "! head -5 data/accession_order.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Everything worked well."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### **Step 7.5. Setting up the color map for visualization in iTOL**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Chiroptera', 'Eulipotyphla', 'Rodentia', 'ND', 'Primates']\n"
     ]
    }
   ],
   "source": [
    "unique_orders = get_unique_orders(\"data/accession_order.txt\")\n",
    "print(unique_orders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Chiroptera': '#32cd32', 'Eulipotyphla': '#ffd700', 'Rodentia': '#1e90ff', 'ND': '#FFFFFF', 'Primates': '#8a2be2'}\n"
     ]
    }
   ],
   "source": [
    "color_map = set_color_map(\"data/accession_order.txt\")\n",
    "print(color_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Colors were set by the user.\n",
      "The request has been fulfilled.\n"
     ]
    }
   ],
   "source": [
    "get_itol_dataset(\"data/accession_organism.txt\", \"data/accession_order.txt\", \"data/dataset_for_iTOL.txt\", color_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 8. Final tree visualization**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Visit [iTOL](https://itol.embl.de)\n",
    "2. Upload `tree/annotated_tree.treefile` as the tree\n",
    "3. Upload `data/dataset_for_iTOL.txt` as the annotation dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **Step 9. Phylogeography**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Предупреждение:\n",
      "пакет 'rnaturalearth' был собран под R версии 4.3.3 \n",
      "\n",
      "Присоединяю пакет: 'dplyr'\n",
      "\n",
      "Следующие объекты скрыты от 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "Следующие объекты скрыты от 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "Предупреждение:\n",
      "пакет 'dplyr' был собран под R версии 4.3.3 \n",
      "Предупреждения:\n",
      "1: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.\n",
      "ℹ Please use `linewidth` instead. \n",
      "2: В geom_segment(aes(x = 39.7015, y = 47.2357, xend = 39.7015, yend = 47.9357 -  :\n",
      "  All aesthetics have length 1, but the data has 39 rows.\n",
      "ℹ Please consider using `annotate()` or provide this layer with data containing\n",
      "  a single row.\n",
      "3: В geom_segment(aes(x = 16.6068, y = 49.1951, xend = 17.75, yend = 49.1951 +  :\n",
      "  All aesthetics have length 1, but the data has 39 rows.\n",
      "ℹ Please consider using `annotate()` or provide this layer with data containing\n",
      "  a single row.\n",
      "4: В st_point_on_surface.sfc(sf::st_zm(x)) :\n",
      "  st_point_on_surface may not give correct results for longitude/latitude data\n",
      "Предупреждения:\n",
      "1: В geom_segment(aes(x = 39.7015, y = 47.2357, xend = 39.7015, yend = 47.9357 -  :\n",
      "  All aesthetics have length 1, but the data has 39 rows.\n",
      "ℹ Please consider using `annotate()` or provide this layer with data containing\n",
      "  a single row.\n",
      "2: В geom_segment(aes(x = 16.6068, y = 49.1951, xend = 17.75, yend = 49.1951 +  :\n",
      "  All aesthetics have length 1, but the data has 39 rows.\n",
      "ℹ Please consider using `annotate()` or provide this layer with data containing\n",
      "  a single row.\n",
      "3: В st_point_on_surface.sfc(sf::st_zm(x)) :\n",
      "  st_point_on_surface may not give correct results for longitude/latitude data\n"
     ]
    }
   ],
   "source": [
    "! Rscript geo-map.R map.png"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "phylo4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
